README.md
Project Summary:
	This project aims to bring a deeper understanding of how sound and reverberation work. This program incorporates tools to calculate the reverberation time of sound in audio files, as well as to help visualize the data gathered from the uploaded files. 

Project Features:

    #1. Data cleaning and conversion:
        Our program will clean any wav file of metadata, as well as convert non-wav files to this format. Furthermore, any multichannel audio will be converted to single channel to improve analysis
    
    #2. Analysis & Visualization
        Our program will take in any audio file the user uploads, and calculate reverberation time as well as visualize it. The waveform of the audio will also be displayed, as well as a spectrogram of the sound.
    
    


Instructions:

	Uploading file: Simply use the 'Select File' button located on top of the program. This will open a file select prompt and ask you to select a .mp3 or .wav file
    
	Data Analysis: The sound file you upload will be handled by the program and the graph of your choosing will be displayed. To choose a graph, simply click one of the corresponding buttons under the name of the selected file.

	Switching Graphs: You will have the ability to change the displayed graph. This can be simply done by clicking the button corresponding to a graph different than that you have selected.


CHANGELOG.md
Changelog:
	v0.1.0:
		Implemented a basic gui
	v0.2.0: 
		Implemented a Reverb.py module to check spectrogram and Power vs. Time
		Implemented a convert.py module to convert mp3 to wav; multichannel to single channel
	v0.2.1:
		Fixed issue where program would not work with different files
	v0.2.2:
		Miscellaneous bug fixes
	v0.3.0:
		Miscellaneous fixes to each module
		Overhauled gui and reverb to be more readable and function properly with one another
		gui.py now displays correct file name
	v0.4.0:
		gui.py now has buttons to display low, mid, and high frequencies
		Fixed issue which prevented gui from displaying frequency graphs correctly
	v0.4.1:
		Fixed issue which caused program to crash when combining graphs
		Fixed issue with how RT60 time displayed
	v0.5.0
		Improved gui and graph display
	v1.0.0
		Finalised audio analysis & visualization
		Fixed Waveform plotting
	v1.0.1
		Updated code for readability/documentation

	
		
		

reverb.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile
from scipy import signal as sp_sig
import wave

# All methods for analysing the file
class Reverb:
    # Initializes variables
    def __init__(self):
        self.sample_rate = None
        self.data = None
        self.spectrum = None
        self.freqs = None
        self.t = None
        self.im = None
        self.data_in_db = None
        self.range = None
        self.RT60 = None

    # Opens the wav file and sets some variables for later use
    def openFile(self, filename):
        self.sample_rate, self.data = wavfile.read(filename)
        self.spectrum, self.freqs, self.t, self.im = plt.specgram(self.data, Fs=self.sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))
        plt.clf()

    # Returns the length of the file in seconds
    def getTime(self):
        return self.t[len(self.t)-1]

    # Takes the frequency range (low, mid, high) and returns an index of the dB values
    def frequency_check(self, ranges):
        index_of_frequency = np.where(self.freqs >= ranges)[0][0]
        data_for_frequency = self.spectrum[index_of_frequency]
        # change a digital signal for a values in decibels
        self.data_in_db = 10 * np.log10(data_for_frequency)

    # find a nearest value
    def find_nearest_value(self, array, value):
        array = np.asarray(array)
        idx = (np.abs(array - value)).argmin()
        return array[idx]

    # Creates the spectrogram graph
    def specgram(self):
        plt.specgram(self.data, Fs=self.sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r'))

    # Creates a dB vs time graph based on the frequency (low, mid, high)
    def dbplt(self, ranges):
        self.frequency_check(ranges)

        # Changes the color of the graph based on the predetermined frequency values
        if 0 < ranges <= 200:
            color = "#0000FF"
        elif 200 < ranges <= 2000:
            color = "#00FF00"
        else:
            color = "#FF0000"

        # Plot dB vs time and label axes
        plt.plot(self.t, self.data_in_db, linewidth=1, alpha=0.7, color=color)
        plt.xlabel('Time (s)')
        plt.ylabel('Power (dB)')

        # Take the frequency (low, mid, high), calculate the RT20, and extrapolate for the RT60
    def getRT60(self, ranges):
        self.frequency_check(ranges)
        index_of_max = np.argmax(self.data_in_db)

        value_of_max = self.data_in_db[index_of_max]

        # slice array from a max value
        sliced_array = self.data_in_db[index_of_max:]

        value_of_max_less_5 = value_of_max - 5
        value_of_max_less_5 = self.find_nearest_value(sliced_array, value_of_max_less_5)

        index_of_max_less_5 = np.where(self.data_in_db == value_of_max_less_5)


        # slice array from a max -5dB
        value_of_max_less_25 = value_of_max - 25

        value_of_max_less_25 = self.find_nearest_value(sliced_array, value_of_max_less_25)

        index_of_max_less_25 = np.where(self.data_in_db == value_of_max_less_25)

        rt20 = (self.t[index_of_max_less_5] - self.t[index_of_max_less_25])[0]

        # extrapolate rt20 to rt60
        self.RT60 = round(abs(3*rt20), 2)

        return self.RT60

    # Returns the frequency (low, mid, high) at which the peak dB value is reached. Used for calculating overall RT60
    def getMaxDBFreq(self):
        self.frequency_check(200)
        low_max_db = self.data_in_db[np.argmax(self.data_in_db)]
        self.frequency_check(1000)
        mid_max_db = self.data_in_db[np.argmax(self.data_in_db)]
        self.frequency_check(4000)
        high_max_db = self.data_in_db[np.argmax(self.data_in_db)]
        if low_max_db > mid_max_db and low_max_db > high_max_db:
            return 200
        elif mid_max_db > high_max_db:
            return 1000
        else:
            return 4000

    # Calls the getRT60() method with the frequency at which the peak dB value occurs, based on the getMaxDBFreq() method
    def getCombinedRT60(self):
        return self.getRT60(self.getMaxDBFreq())

    # Uses welch to find the resonant frequency
    def getResonantFreq(self):
        frequencies, power = sp_sig.welch(self.data, self.sample_rate, nperseg=4096)
        dominant_frequency = frequencies[np.argmax(power)]
        return round(dominant_frequency)

    # Plots the waveform of the sound file
    def waveformplt(self):
        waveform = wave.open("pt_mono.wav", "r")
        raw = waveform.readframes(-1)
        raw = np.frombuffer(raw, "int16")
        plt.title("Waveform")
        plt.plot(raw, color="#FF9999")
        

gui.py
import os.path
from tkinter import *
from tkinter import ttk     # from tkinter import Tk for Python 3.x
from tkinter.filedialog import askopenfilename
from convert import *
from reverb import *

# Initialize class variable to call methods from
reverb = Reverb()

# Handles getting the file from user and calls conversion functions
def selectFile():
    filename = askopenfilename(filetypes=(("wav File", "*.wav"), ("mp3 File", "*.mp3")))
    if filename.endswith(".wav"):
        toMono(filename)
    elif filename.endswith(".mp3"):
        mp3towav(filename)
        toMono("pt_mono.wav")
    else:
        filename_label.config(text="Please choose a .mp3 or .wav file.")
        return

    # Enables graph buttons when acceptable file is selected
    lowRT60Btn.config(state="normal")
    midRT60Btn.config(state="normal")
    highRT60Btn.config(state="normal")
    combRT60Btn.config(state="normal")
    waveformBtn.config(state="normal")
    spectBtn.config(state="normal")

    # Opens the file
    reverb.openFile("pt_mono.wav")

    # Displays the name, length, resonant frequency, and RT60 of the selected file
    filename_label.config(text=f"Current file: {os.path.basename(filename)}")
    time_label.config(text=f"{round(reverb.getTime(), 2)} seconds.")
    ResonantFrequency_label.config(text=f"Resonant frequency: {reverb.getResonantFreq()} Hz")
    updateRT60(reverb.getCombinedRT60())

# Used to change the RT60 value displayed in the tkinter window
def updateRT60(RT60):
    RT60_label.config(text=f"The RT60 is {RT60} seconds.")
    RT60diff_label.config(text=f"The RT60 diff is {round(RT60-0.5, 2)}")

# Tkinter stuff
root = Tk()
root.title('S. P. I. D. A. M.')
root.geometry("200x278")

_mainframe = ttk.Frame(root, padding='5 5 5 5 ')
_mainframe.grid(row=0, column=0, sticky=("E", "W", "N", "S"))

# All labels
textFrame = ttk.Frame(root, relief='sunken')
textFrame.grid(row = 2, column = 0, sticky=("E", "W", "S"))
textFrame = ttk.LabelFrame(_mainframe, text = '')

filename_label = ttk.Label(_mainframe, text = '')
filename_label.grid(row = 2, column = 0)

time_label = ttk.Label(_mainframe, text = '')
time_label.grid(row = 3, column = 0)

RT60_label = ttk.Label(_mainframe, text = '')
RT60_label.grid(row = 4, column = 0)

RT60diff_label = ttk.Label(_mainframe, text = '')
RT60diff_label.grid(row = 5, column = 0)

ResonantFrequency_label = ttk.Label(_mainframe, text = '')
ResonantFrequency_label.grid(row = 6, column = 0)

# All buttons
selectBtn = ttk.Button(_mainframe, text='Select File', command=selectFile)
selectBtn.grid(row=1, column = 0, sticky = N, padx = 55)

lowRT60Btn = ttk.Button(_mainframe, state="disabled", text='Low RT60', command=lambda: (updateRT60(reverb.getRT60(200)), plt.clf(), reverb.dbplt(200), plt.show()))
lowRT60Btn.grid(row=7, column = 0, sticky = N, padx = 10)

midRT60Btn = ttk.Button(_mainframe, state="disabled", text='Mid RT60', command=lambda: (updateRT60(reverb.getRT60(1000)), plt.clf(), reverb.dbplt(1000), plt.show()))
midRT60Btn.grid(row=8, column = 0, sticky = N, padx = 10)

highRT60Btn = ttk.Button(_mainframe, state="disabled", text='High RT60', command=lambda: (updateRT60(reverb.getRT60(4000)), plt.clf(), reverb.dbplt(4000), plt.show()))
highRT60Btn.grid(row=9, column = 0, sticky = N, padx = 10)

combRT60Btn = ttk.Button(_mainframe, state="disabled", text='Combined RT60', command=lambda: (updateRT60(reverb.getCombinedRT60()), plt.clf(), reverb.dbplt(200), reverb.dbplt(1000), reverb.dbplt(4000), plt.show()))
combRT60Btn.grid(row=10, column = 0, sticky = N, padx = 10)

waveformBtn = ttk.Button(_mainframe, state="disabled", text='Waveform', command=lambda: (updateRT60(reverb.getCombinedRT60()), plt.clf(), reverb.waveformplt(), plt.show()))
waveformBtn.grid(row=11, column = 0, sticky = N, padx = 10)

spectBtn = ttk.Button(_mainframe, state="disabled", text='Spectrogram', command=lambda: (updateRT60(reverb.getCombinedRT60()), plt.clf(), reverb.specgram(), plt.show()))
spectBtn.grid(row=12, column = 0, sticky = N, padx = 10)

# Only run if main
if __name__ == "__main__":
    root.mainloop()
